common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java

```diff
@@ -15,53 +15,75 @@
  */
 package io.netty.util.concurrent;
 
+import io.netty.util.internal.logging.InternalLogger;
+
 import java.util.Arrays;
+import java.util.EventListener;
 
 final class DefaultFutureListeners {
-
-    private GenericFutureListener<? extends Future<?>>[] listeners;
+    private Object[] listeners;
     private int size;
 
-    @SuppressWarnings("unchecked")
-    DefaultFutureListeners(
-            GenericFutureListener<? extends Future<?>> first, GenericFutureListener<? extends Future<?>> second) {
-        listeners = new GenericFutureListener[2];
-        listeners[0] = first;
-        listeners[1] = second;
-        size = 2;
+    DefaultFutureListeners() {
+        listeners = new Object[4];
     }
 
-    public void add(GenericFutureListener<? extends Future<?>> l) {
-        GenericFutureListener<? extends Future<?>>[] listeners = this.listeners;
-        final int size = this.size;
-        if (size == listeners.length) {
+    public void add(EventListener listener, Object context) {
```

Comment:
We might consider requiring `context` to be non-null for `FutureContextListener`, so we can reliably determine which of the `addListener()` methods were called by whether context is null within `listeners`. _Technically_ an object could be both a `FutureContextListener` and a `FutureListener`.

Or probably better, could have a `Object NULL_CONTEXT = new Object()` that we use for `context` if it happens to be `null`, and then use `null` to imply `FutureListener`.